openfasoc/generators/gdsfactory-gen/pygen/routing/c_route.py

```diff
@@ -0,0 +1,208 @@
+	e2_length = extension + evaluate_bbox(viastack2)[0]
+	xdiff = abs(edge1.center[0] - edge2.center[0])
+	ydiff = abs(edge1.center[1] - edge2.center[1])
+	if not isclose(edge1.center[0],edge2.center[0]):
+		if round(edge1.orientation) == 0:# facing east
+			if edge1.center[0] > edge2.center[0]:
+				e2_length += xdiff
+			else:
+				e1_length += xdiff
+		elif round(edge1.orientation) == 180:# facing west
+			if edge1.center[0] < edge2.center[0]:
+				e2_length += xdiff
+			else:
+				e1_length += xdiff
+	if not isclose(edge1.center[1],edge2.center[1]):
+		if round(edge1.orientation) == 270:# facing south
+			if edge1.center[1] < edge2.center[1]:
+				e2_length += ydiff
+			else:
+				e1_length += ydiff
+		elif round(edge1.orientation) == 90:#facing north
+			if edge1.center[1] > edge2.center[1]:
+				e2_length += ydiff
+			else:
+				e1_length += ydiff
+	# move into position
+	e1_extension_comp = Component("edge1 extension")
+	e2_extension_comp = Component("edge2 extension")
+	box_dims = [(e1_length, width1),(e2_length, width2)]
+	if round(edge1.orientation) == 90 or round(edge1.orientation) == 270:
+		box_dims = [(width1, e1_length),(width2, e2_length)]
+	rect_c1 = rectangle(size=box_dims[0], layer=pdk.get_glayer(e1glayer),centered=True).copy()
+	rect_c2 = rectangle(size=box_dims[1], layer=pdk.get_glayer(e2glayer),centered=True).copy()
+	rect_c1 = rename_ports_by_orientation(rename_ports_by_list(rect_c1,[("e","e_")]))
+	rect_c2 = rename_ports_by_orientation(rename_ports_by_list(rect_c2,[("e","e_")]))
+	e1_extension = e1_extension_comp << rect_c1
+	e2_extension = e2_extension_comp << rect_c2
+	e1_extension.move(destination=edge1.center)
+	e2_extension.move(destination=edge2.center)
+	if round(edge1.orientation) == 0:# facing east
+		e1_extension.movex(evaluate_bbox(e1_extension)[0]/2)
+		e2_extension.movex(evaluate_bbox(e2_extension)[0]/2)
+	elif round(edge1.orientation) == 180:# facing west
+		e1_extension.movex(0-evaluate_bbox(e1_extension)[0]/2)
+		e2_extension.movex(0-evaluate_bbox(e2_extension)[0]/2)
+	elif round(edge1.orientation) == 270:# facing south
+		e1_extension.movey(0-evaluate_bbox(e1_extension)[1]/2)
+		e2_extension.movey(0-evaluate_bbox(e2_extension)[1]/2)
+	else:#facing north
+		e1_extension.movey(evaluate_bbox(e1_extension)[1]/2)
+		e2_extension.movey(evaluate_bbox(e2_extension)[1]/2)
+	# place viastacks
+	e1_extension_comp.add_ports(e1_extension.get_ports_list())
+	e2_extension_comp.add_ports(e2_extension.get_ports_list())
+	me1 = e1_extension_comp << viastack1
+	me2 = e2_extension_comp << viastack2
+	route_ports = [None,None]
+	via_flush = abs((width1 - evaluate_bbox(viastack1)[0])/2) if viaoffset else 0
+	via_flush1 = via_flush if viaoffset[0] else 0-via_flush
+	via_flush1 = 0 if viaoffset[0] is None else via_flush1
+	via_flush2 = via_flush if viaoffset[1] else 0-via_flush
+	via_flush2 = 0 if viaoffset[1] is None else via_flush2
+	if round(edge1.orientation) == 0:# facing east
```

Comment:
It seems like the code under this function is repeated many times. Is there a possibility for refactoring?